geo/src/algorithm/triangulate_spade.rs

```diff
@@ -0,0 +1,710 @@
+                })
+                .collect::<Vec<_>>()
+        })
+    }
+}
+
+/// conversion from spade triangulation back to geo triangles
+fn triangulation_to_triangles<T, F>(triangulation: T) -> Triangles<F>
+where
+    T: Triangulation<Vertex = Point2<F>>,
+    F: SpadeTriangulationFloat,
+{
+    triangulation
+        .inner_faces()
+        .map(|face| face.positions())
+        .map(|points| points.map(|p| Coord::<F> { x: p.x, y: p.y }))
+        .map(Triangle::from)
+        .collect::<Vec<_>>()
+}
+
+// ========== Triangulation trait impls ============
+
+// everything that satisfies the requirement methods automatically implements the triangulation
+impl<'a, T, G> TriangulateSpade<'a, T> for G
+where
+    T: SpadeTriangulationFloat,
+    G: private::TriangulationRequirementTrait<'a, T>,
+{
+}
+
+impl<'a, 'l, T, G> private::TriangulationRequirementTrait<'a, T> for G
+where
+    'a: 'l,
+    T: SpadeTriangulationFloat,
+    G: LinesIter<'l, Scalar = T> + CoordsIter<Scalar = T> + Contains<Point<T>>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.coords_iter().collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.lines_iter().collect()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.contains(&p)
+    }
+}
+
+// it would be cool to impl the trait for GS: AsRef<[G]> but I wasn't able to get this to compile
+// (yet)
+
+impl<'a, T, G> private::TriangulationRequirementTrait<'a, T> for Vec<G>
+where
+    T: SpadeTriangulationFloat,
+    G: TriangulateSpade<'a, T>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.iter().flat_map(|g| g.get_points()).collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.iter().flat_map(|g| g.get_lines()).collect::<Vec<_>>()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.iter().any(|g| g.contains_point(p))
+    }
+}
+
+impl<'a, T, G> private::TriangulationRequirementTrait<'a, T> for &[G]
+where
+    T: SpadeTriangulationFloat,
+    G: TriangulateSpade<'a, T>,
+{
+    fn get_points(&'a self) -> Vec<Coord<T>> {
+        self.iter().flat_map(|g| g.get_points()).collect::<Vec<_>>()
+    }
+
+    fn get_lines(&'a self) -> Vec<Line<T>> {
+        self.iter().flat_map(|g| g.get_lines()).collect::<Vec<_>>()
+    }
+
+    fn contains_point(&'a self, p: Point<T>) -> bool {
+        self.iter().any(|g| g.contains_point(p))
+    }
+}
+
+// ========== Triangulation trait impl helpers ============
+
+fn prepare_intersection_contraint<T: SpadeTriangulationFloat>(
+    mut lines: Vec<Line<T>>,
+    mut known_points: Vec<Coord<T>>,
+) -> Result<Vec<Line<T>>, TriangulationError> {
+    // safety net. We can't prove that the `while let` loop isn't going to run infinitely, so
+    // we abort after a fixed amount of iterations
+    let mut loop_count = 1000;
+    // in case of an error we have something to return (the scenario that triggered the infinite
```

Comment:
I'm not sure what's meant by this comment. Where do we return the scenario that triggered the infinite loop?

Reply:
Yeah, that is just poor wording. I was just trying to implement [Rule 2 of "The Power of 10" rules of NASA](https://en.wikipedia.org/wiki/The_Power_of_10%3A_Rules_for_Developing_Safety-Critical_Code). It just returns the `LoopTrap` Error to prevent indefinite loops although that didn't occur yet. But you never know 🤷🏼